modules/calib3d/doc/solvePnP.markdown

```diff
@@ -0,0 +1,176 @@
+-   cv::SOLVEPNP_IPPE Method is based on the paper of T. Collins and A. Bartoli.
+"Infinitesimal Plane-Based Pose Estimation" (@cite Collins14). This method requires coplanar object points.
+-   cv::SOLVEPNP_IPPE_SQUARE Method is based on the paper of Toby Collins and Adrien Bartoli.
+"Infinitesimal Plane-Based Pose Estimation" (@cite Collins14). This method is suitable for marker pose estimation.
+It requires 4 coplanar object points defined in the following order:
+  - point 0: [-squareLength / 2,  squareLength / 2, 0]
+  - point 1: [ squareLength / 2,  squareLength / 2, 0]
+  - point 2: [ squareLength / 2, -squareLength / 2, 0]
+  - point 3: [-squareLength / 2, -squareLength / 2, 0]
+-   cv::SOLVEPNP_SQPNP Method is based on the paper "A Consistently Fast and Globally Optimal Solution to the
+Perspective-n-Point Problem" by G. Terzakis and M.Lourakis (@cite Terzakis2020SQPnP). It requires 3 or more points.
+
+## P3P
+
+The cv::solveP3P() computes an object pose from exactly 3 3D-2D point correspondences. A P3P problem has up to 4 solutions.
+
+@note The solutions are sorted by reprojection errors (lowest to highest).
+
+## PnP
+
+The cv::solvePnP() returns the rotation and the translation vectors that transform a 3D point expressed in the object
+coordinate frame to the camera coordinate frame, using different methods:
+- P3P methods (cv::SOLVEPNP_P3P, cv::SOLVEPNP_AP3P): need 4 input points to return a unique solution.
+- cv::SOLVEPNP_IPPE Input points must be >= 4 and object points must be coplanar.
+- cv::SOLVEPNP_IPPE_SQUARE Special case suitable for marker pose estimation.
+Number of input points must be 4. Object points must be defined in the following order:
+  - point 0: [-squareLength / 2,  squareLength / 2, 0]
+  - point 1: [ squareLength / 2,  squareLength / 2, 0]
+  - point 2: [ squareLength / 2, -squareLength / 2, 0]
+  - point 3: [-squareLength / 2, -squareLength / 2, 0]
+- for all the other flags, number of input points must be >= 4 and object points can be in any configuration.
+
+## Generic PnP
+
+The cv::solvePnPGeneric() allows retrieving all the possible solutions.
+
+Currently, only cv::SOLVEPNP_P3P, cv::SOLVEPNP_AP3P, cv::SOLVEPNP_IPPE, cv::SOLVEPNP_IPPE_SQUARE, cv::SOLVEPNP_SQPNP can return multiple solutions.
+
+## RANSAC PnP
+
+The cv::solvePnPRansac() computes the object pose wrt. the camera frame using a RANSAC scheme to deal with outliers.
+
+More information can be found in @cite Zuliani2014RANSACFD
+
+## Pose refinement
+
+Pose refinement consists in estimating the rotation and translation that minimizes the reprojection error using a non-linear minimization method and starting from an initial estimate of the solution. OpenCV proposes cv::solvePnPRefineLM() and cv::solvePnPRefineVVS() for this problem.
+
+cv::solvePnPRefineLM() uses a non-linear Levenberg-Marquardt minimization scheme @cite Madsen04 @cite Eade13 and the current implementation computes the rotation update as a perturbation and not on SO(3).
```

Comment:
@savuor 

Is it correct for the current implementation that the rotation update is not done on SO(3)? So it should only be valid for small rotation perturbation.

What about the future LM implementation?

Reply:
@catree 

I thought about it when reimplementing LevMarq. Of course, it's better to do it in true way with SO(3) and exponential oplus but:
0. The current code works
1. AFAIK the difference between the true method and the existing method has more influence on convergence speed than on result correctness
2. LevMarq is a local optimization algorithm, it doesn't have to work well on large perturbations

The ongoing LevMarq implementation will let a user (OK, maybe not an OpenCV user but at least an OpenCV contributor) to implement objective functions which accept SE(3) among other things.

Reply:
Thanks for your answer. This makes sense.

> (1) AFAIK the difference between the true method and the existing method has more influence on a convergence speed than on a result correctness

Maybe some related discussion: https://forum.opencv.org/t/slow-convergence-of-solvepnprefinelm/5486

---

Wondering if the new implementation would let adding additional constraints such as line of sight or similar ideas into the solving framework?

Reply:
Thanks for the link to the discussion. Now it's obvious to me that sometimes bad convergence speed means bad accuracy.

I'm not sure what exactly this line of sight constraint is. If it can be done as a part of cost function in a regularization manner then even old LevMarq is able to work with them.

Reply:
Thanks for the info and the work done 👍 

---

On another topic, there is this paper: [A compact formula for the derivative of a 3-D rotation in exponential coordinates](https://arxiv.org/pdf/1312.0788.pdf). Not sure if it can be useful pratically.

Reply:
Regarding the 3D rotation derivative: usually an exponential derivative near 0 is used in local optimization, see [this example](https://github.com/opencv/opencv/blob/5.x/modules/3d/src/rgbd/pose_graph.cpp#L162).